### PR TITLE
[LLHD] Add extract field operations

### PR DESCRIPTION
**Pull request by maerhart**
* Add operations to statically and dynamically extract an element from
a vector or tuple
* They can also operate on signals by returning an alias signal to the
specified element of the underlying type of the target signal, this
means they don't perform any kind of probing like the llhd.prb
instruction

### CODE REVIEW

File: include/circt/Dialect/LLHD/IR/ExtractOps.td [hunk 1]
```diff
@@ -101,3 +101,118 @@ def LLHD_DextsOp : LLHD_Op<"dexts", [
     unsigned getTargetWidth() { return getLLHDTypeWidth(target().getType()); }
   }];
 }
+
+def LLHD_ExtfOp : LLHD_Op<"extf", [
```
Comment (stephenneuendorffer) [inline include/circt/Dialect/LLHD/IR/ExtractOps.td:105]: prefer complete words.  IS this "Extend Float?" "Extract field?" "External foobar?" "Example Tensorflow?"

File: include/circt/Dialect/LLHD/IR/ExtractOps.td [hunk 2]
```diff
@@ -101,3 +101,118 @@ def LLHD_DextsOp : LLHD_Op<"dexts", [
...
+    NoSideEffect,
+    PredOpTrait<"'index' has to be smaller than the 'target' size",
+      CPred<"$index.cast<IntegerAttr>().getInt() < getTargetWidth()">>,
+    TypesMatchWith<"'result' type has to match type at 'index' of 'target', in "
+      "case 'target' is a singal, consider the underlying types of the 'target'"
```
Comment (stephenneuendorffer) [inline include/circt/Dialect/LLHD/IR/ExtractOps.td:110]: singal -> signal

File: include/circt/Dialect/LLHD/IR/ExtractOps.td [hunk 3]
```diff
@@ -101,3 +101,118 @@ def LLHD_DextsOp : LLHD_Op<"dexts", [
     unsigned getTargetWidth() { return getLLHDTypeWidth(target().getType()); }
   }];
 }
+
+def LLHD_ExtfOp : LLHD_Op<"extf", [
+    NoSideEffect,
+    PredOpTrait<"'index' has to be smaller than the 'target' size",
+      CPred<"$index.cast<IntegerAttr>().getInt() < getTargetWidth()">>,
+    TypesMatchWith<"'result' type has to match type at 'index' of 'target', in "
```
Comment (stephenneuendorffer) [inline include/circt/Dialect/LLHD/IR/ExtractOps.td:109]: This seems awkwardly worded to me.  The description doesn't quite seem to match the predicate?

File: include/circt/Dialect/LLHD/IR/ExtractOps.td [hunk 4]
```diff
@@ -101,3 +101,118 @@ def LLHD_DextsOp : LLHD_Op<"dexts", [
...
+      "case 'target' is a singal, consider the underlying types of the 'target'"
+      " and 'result' signals",
+      "target", "result",
+      "($_self.isa<llhd::SigType>() "
+        "? llhd::SigType::get("
+          "getElementTypeAtIndex($index.cast<IntegerAttr>().getInt())) "
+        ": getElementTypeAtIndex($index.cast<IntegerAttr>().getInt()))">
+  ]> {
+  let summary = "Extract an element from a vector or tuple";
```
Comment (stephenneuendorffer) [inline include/circt/Dialect/LLHD/IR/ExtractOps.td:118]: or a signal with with vector or tuple type?

File: include/circt/Dialect/LLHD/IR/ExtractOps.td [hunk 5]
```diff
@@ -101,3 +101,118 @@ def LLHD_DextsOp : LLHD_Op<"dexts", [
...
+      if (auto sig = targetType.dyn_cast<llhd::SigType>())
+        targetType = sig.getUnderlyingType();
+      if (auto vec = targetType.dyn_cast<VectorType>())
+        return vec.getElementType();
+      return targetType.dyn_cast<TupleType>().getTypes()[index];
+    }
+  }];
+}
+
+def LLHD_DextfOp : LLHD_Op<"dextf", [
```
Comment (stephenneuendorffer) [inline include/circt/Dialect/LLHD/IR/ExtractOps.td:167]: Oh no, it's random sequence of letters op!

**Comment by maerhart** [comment]
@stephenneuendorffer Thanks for your review! I renamed `llhd.extf -> llhd.extract_element` and `llhd.dextf -> llhd.dyn_extract_element`.  
I'll also rename the already existing insert and extract operations in another PR as follows:
`llhd.exts -> llhd.extract_slice`
`llhd.dexts -> llhd.dyn_extract_slice`
`llhd.inss -> llhd.insert_slice`
`llhd.insf -> llhd.insert_element`
Or does someone have a suggestion for another (better) naming scheme?

**Comment by rodonisi** [comment]
> @stephenneuendorffer Thanks for your review! I renamed `llhd.extf -> llhd.extract_element` and `llhd.dextf -> llhd.dyn_extract_element`.
> I'll also rename the already existing insert and extract operations in another PR as follows:
> `llhd.exts -> llhd.extract_slice`
> `llhd.dexts -> llhd.dyn_extract_slice`
> `llhd.inss -> llhd.insert_slice`
> `llhd.insf -> llhd.insert_element`
> Or does someone have a suggestion for another (better) naming scheme?

Sounds good I think.

**Pull request merged**